src/lib/meta/meta_blk_service.cpp

```diff
@@ -1074,84 +1082,109 @@ sisl::byte_array MetaBlkService::read_sub_sb_internal(const meta_blk* mblk) cons
 void MetaBlkService::recover(bool do_comp_cb) {
     // for each registered subsystem, look up in cache for their meta blks;
     std::lock_guard< decltype(m_shutdown_mtx) > lg{m_shutdown_mtx};
-    for (auto& m : m_meta_blks) {
-        auto* mblk = m.second;
-        auto buf = read_sub_sb_internal(mblk);
-
-        // found a meta blk and callback to sub system;
-        const auto itr = m_sub_info.find(mblk->hdr.h.type);
-        if (itr != std::end(m_sub_info)) {
-            // if subsystem registered crc protection, verify crc before sending to subsystem;
-            if (itr->second.do_crc) {
-                const auto crc = crc32_ieee(init_crc32, s_cast< const uint8_t* >(buf->bytes), mblk->hdr.h.context_sz);
-
-                HS_REL_ASSERT_EQ(crc, uint32_cast(mblk->hdr.h.crc),
-                                 "[type={}], CRC mismatch: {}/{}, on mblk bid: {}, context_sz: {}", mblk->hdr.h.type,
-                                 crc, uint32_cast(mblk->hdr.h.crc), mblk->hdr.h.bid.to_string(),
-                                 uint64_cast(mblk->hdr.h.context_sz));
-            } else {
-                HS_LOG(DEBUG, metablk, "[type={}] meta blk found with bypassing crc.", mblk->hdr.h.type);
+    meta_subtype_vec_t ordered_subtypes;
+
+    if (hs_utils::topological_sort(m_dep_topo_graph, ordered_subtypes)) {
+        throw homestore::homestore_exception(
+            "MetaBlkService has circular dependency, please check the dependency graph", homestore_error::init_failed);
+    }
+
+    // all the subsystems are divided into two parts.
+    // for subsystems in ordered_subtypes, we need to recover in order.
+    for (const auto& subtype : ordered_subtypes) {
```

Comment:
recovery_meta_block should take client type and move line: 1095~ 1101, 1110 ~1118 to a new recover_meta_block(client_type) which takes client type as input. This function now provides recovery (return all metablks) for this client type, not just for a given metablk cookie. 

```
for (const auto& subtype : ordered_subtypes) {
    recover_meta_blk(subtype);
}

 for (auto const& x : m_sub_info) {
        auto& reg_info = x.second;
        if (!reg_info.has_deps) {
            recover_meta_block(x.first);
        }
```

```
void recover_meta_block(meta_sub_type client) {
        for (const auto& m : meta_sub_info[client].meta_bids) {
                auto mblk = m_meta_blks[m];
                recover_meta_block(mblk);
            }

            if (do_comp_cb && reg_info.comp_cb) {
                reg_info.comp_cb(true);
                HS_LOG(DEBUG, metablk, "[type={}] completion callback sent.", x.first);
            }
}
```

Reply:
ack